src/realm/object-store/shared_realm.cpp

```diff
@@ -912,6 +913,10 @@ void Realm::close()
 
 void Realm::delete_files(const std::string& realm_file_path)
 {
+    if (!realm::util::File::exists(realm_file_path)) {
```

Comment:
Don't you need this check to be inside the "lock_successful" block below to protect against a race?

Reply:
If `realm_file_path` doesn't exist, `DB::call_with_lock` will fail. In fact, it will only fail if the folder doesn't exist, but will succeed if the folder exists, but the lock file itself doesn't.

Reply:
@finnschiermer `DB::call_with_lock` will create the file, if it is not there.

Reply:
@finnschiermer What @jedelbo said. :)

https://github.com/realm/realm-core/blob/master/src/realm/db.cpp#L2399:
```
lockfile.open(lockfile_path, File::access_ReadWrite, File::create_Auto, 0); // Throws
```

Reply:
Assume two concurrent calls to realm::delete_files. Since the realm file exists, both calls passes the check and enter DB::call_with_lock(). The first one gets the lock and does the callback. The callback removes the management directory (and the files), then the second one gets to run, gets the exclusive lock (since the previous one has released it), and does its callback. This time the directory does not exist.

Reply:
I believe that will be fine as all the deletions inside `call_with_lock` are `try_delete_xxx`, so if the first `delete_files` removed everything, the second will just do nothing. The more problematic case is if I call `delete_files("/usr/non-existent-folder/my.realm")`. `call_with_lock` will try to create the lock in a folder that doesn't exist and throw an exception. But if I call `delete_files("/usr/existing-folder/my.realm")`, even if `my.realm` and its lock file don't exist, `delete_files` will happily succeed. This inconsistent behavior is going to be annoying for developers.

Of course, there are alternatives ways to solve this:
1. We can make `lockfile.open` create the entire folder structure before creating the lock.
2. We can make `call_with_lock` return `false` if the lock doesn't exist

Those are more involved changes though, and I'm not sure if it's worth making them.

Reply:
Ah, Ok, so you're fine with try_remove() throwing an exception.

Reply:
With respect to the directory holding the lock file... we've never tried to do anything to that (add it, delete it, whatever) so that'd be a new requirement.